### PR TITLE
Add long answer candidates to natural questions dataset

### PR DESCRIPTION
**Pull request by seirasto**
This is a modification of the Natural Questions dataset to include missing information specifically related to long answer candidates. (See here: https://github.com/google-research-datasets/natural-questions#long-answer-candidates). This information is important to ensure consistent comparison with prior work. It does not disturb the rest of the format . @lhoestq @albertvillanova

### CODE REVIEW

File: datasets/natural_questions/README.md [hunk 1]
```diff
@@ -115,6 +115,12 @@ The data fields are the same among all splits.
     - `end_byte`: a `int64` feature.
     - `text`: a `string` feature.
   - `yes_no_answer`: a classification label, with possible values including `NO` (0), `YES` (1).
+  - `long_answer_candidates`: a dictionary feature containing:
+   - `start_token`: a `int64` feature.
+   - `end_token`: a `int64` feature.
+   - `start_byte`: a `int64` feature.
+   - `end_byte`: a `int64` feature.
+   - `top_level`: a `bool` feature.
```
Comment (albertvillanova) [inline datasets/natural_questions/README.md:123]: Two comments in relation to this:
- The 5 lines below `long_answer_candidates` should be indented an additional blank space: Markdown sub-lists need at least 2-blanks indent; this is why they are not properly rendered here: https://github.com/huggingface/datasets/blob/fdc87c43a9a24fd2098fb28ebe9abee187fda52d/datasets/natural_questions/README.md
- For consistency, could you please add these lines above `annotations`?

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thanks a lot @seirasto for your contribution. Definitely, the addition of this field will be useful for all community members using this dataset.

Just some comments below.

In relation with the non-passing tests, please note that once the changes are validated, we should pre-process all this dataset and upload it to our HuggigFace cloud.

File: datasets/natural_questions/README.md (hunk 1 above)
Comment (seirasto) [reply datasets/natural_questions/README.md]: I've fixed the README. Do I need to do something for the non-passing tests? I believe these are connected to the pre-processing of the dataset which you will take care of.

Comment (albertvillanova) [reply datasets/natural_questions/README.md]: Thanks.

**Comment by albertvillanova** [comment]
Once we have added `long_answer_candidates` maybe it would be worth to also add the missing `candidate_index` (inside `long_answer`). What do you think, @seirasto ?

**Comment by seirasto** [comment]
@albertvillanova I've added the missing fields and updated the README to include a data instance and some other things.

**Comment by seirasto** [comment]
Great! I've made the updates to align the README. Please let me know if I missed anything.

**Comment by albertvillanova** [comment]
As there were many minor little fixes, I thought it would be easier to fix them directly.

**Comment by lhoestq** [comment]
It looks good to me, thanks @seirasto !

**Comment by albertvillanova** [comment]
I have merged the master branch, so that we include all the fixes on Apache Beam + Google Dataflow.

**Comment by albertvillanova** [comment]
Pre-processing is running!

Already finished for "dev" config:
```python
In [2]: ds = load_dataset("datasets/natural_questions", "dev")

In [3]: ds
Out[3]: 
DatasetDict({
    validation: Dataset({
        features: ['id', 'document', 'question', 'long_answer_candidates', 'annotations'],
        num_rows: 7830
    })
})
```

**Comment by albertvillanova** [comment]
There is an issue while running the preprocessing for the "default" (train+dev) config. Train data files are larger than than dev ones and workers run out of memory.

I'm opening a separate issue to handle this problem:

**Comment by albertvillanova** [comment]
@seirasto is proposing uploading their preprocessed data files to our Datasets bucket.

I think @lhoestq can give a more informed answer about authentication requirements.

**Comment by lhoestq** [comment]
Now that the data fiels are uploaded, can you merge the `main` branch into yours to re-trigger the CI @seirasto please ? :) Then I think we can merge if it's good for you @albertvillanova

**Review by albertvillanova — APPROVED** [review]
Good, once CI is green! Finally we can merge this PR... :sweat_smile:

**Comment by seirasto** [comment]
Merge is done! I think someone needs to approve the CI to run :)

**Comment by lhoestq** [comment]
Can you run `make style` to fix the code formatting required by the CI please ?

File: datasets/natural_questions/natural_questions.py [hunk 2]
```diff
@@ -74,13 +74,23 @@ def _info(self):
                         "url": datasets.Value("string"),
                         "html": datasets.Value("string"),
                         "tokens": datasets.features.Sequence(
-                            {"token": datasets.Value("string"), "is_html": datasets.Value("bool")}
+                            {"token": datasets.Value("string"), "is_html": datasets.Value("bool"),  
+                            "start_byte": datasets.Value("int64"), "end_byte": datasets.Value("int64")}
```
Comment (albertvillanova) [inline datasets/natural_questions/natural_questions.py:78]:
```suggestion
                            {
                                "token": datasets.Value("string"),
                                "is_html": datasets.Value("bool"),
                                "start_byte": datasets.Value("int64"),
                                "end_byte": datasets.Value("int64"),
                            }
```

File: datasets/natural_questions/natural_questions.py [hunk 3]
```diff
@@ -89,6 +99,7 @@ def _info(self):
                                 "end_token": datasets.Value("int64"),
                                 "start_byte": datasets.Value("int64"),
                                 "end_byte": datasets.Value("int64"),
+                                "candidate_index": datasets.Value("int64")
```
Comment (albertvillanova) [inline datasets/natural_questions/natural_questions.py:102]:
```suggestion
                                "candidate_index": datasets.Value("int64"),
```

File: datasets/natural_questions/natural_questions.py [hunk 4]
```diff
@@ -162,6 +173,7 @@ def _parse_annotation(an_json):
                         "end_token": an_json["long_answer"]["end_token"],
                         "start_byte": an_json["long_answer"]["start_byte"],
                         "end_byte": an_json["long_answer"]["end_byte"],
+                        "candidate_index": an_json["long_answer"]["candidate_index"]
```
Comment (albertvillanova) [inline datasets/natural_questions/natural_questions.py:176]:
```suggestion
                        "candidate_index": an_json["long_answer"]["candidate_index"],
```

File: datasets/natural_questions/natural_questions.py [hunk 5]
```diff
@@ -179,10 +191,11 @@ def _parse_annotation(an_json):
                         "url": ex_json["document_url"],
                         "html": ex_json["document_html"],
                         "tokens": [
-                            {"token": t["token"], "is_html": t["html_token"]} for t in ex_json["document_tokens"]
+                            {"token": t["token"], "is_html": t["html_token"], "start_byte": t["start_byte"], "end_byte": t["end_byte"]} for t in ex_json["document_tokens"]
```
Comment (albertvillanova) [inline datasets/natural_questions/natural_questions.py:194]:
```suggestion
                            {
                                "token": t["token"],
                                "is_html": t["html_token"],
                                "start_byte": t["start_byte"],
                                "end_byte": t["end_byte"],
                            }
                            for t in ex_json["document_tokens"]
```

**Review by albertvillanova — COMMENTED** [review]
After running `make style`, these are the style fixes (see below).

**Comment by seirasto** [comment]
Thanks @albertvillanova! I've committed all your suggestions.

**Comment by albertvillanova** [comment]
The CI is green. I'm merging this PR.

**Pull request merged**